File: VB/PivotGridDrillDown.Module.Web/Controllers/PivotDrillDownController.vb
Imports System
Imports System.Collections
Imports System.Linq
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.PivotGrid.Web
Imports DevExpress.ExpressApp.Web
Imports DevExpress.ExpressApp.Web.Templates
Imports DevExpress.ExpressApp.Web.Utils
Imports DevExpress.Web.ASPxPivotGrid
Imports DevExpress.XtraPivotGrid

Namespace PivotGridDrillDown.Module.Web.Controllers
    Public Class PivotDrillDownController
        Inherits ViewController(Of ListView)
        Implements IXafCallbackHandler
        Private ReadOnly handlerId As String
        Protected ReadOnly Property CallbackManager() As XafCallbackManager
            Get
                Return DirectCast(WebWindow.CurrentRequestPage, ICallbackManagerHolder).CallbackManager
            End Get
        End Property
        Protected Overrides Sub OnViewControlsCreated()
            MyBase.OnViewControlsCreated()

            CallbackManager.RegisterHandler(handlerId, Me)

            Dim pivotGridListEditor As ASPxPivotGridListEditor = TryCast(View.Editor, ASPxPivotGridListEditor)
            If pivotGridListEditor IsNot Nothing Then
                Dim pivotGrid As ASPxPivotGrid = pivotGridListEditor.PivotGridControl
                Dim script As String = CallbackManager.GetScript(handlerId, "e.ColumnIndex + ';' + e.RowIndex")
                ClientSideEventsHelper.AssignClientHandlerSafe(pivotGrid, "CellClick", (Convert.ToString("function(s, e) {") & script) + "}", "pivotGrid.CellClick")
            End If
        End Sub
        Public Sub New()
            handlerId = "PivotDrillDownHandler" + GetHashCode().ToString()
        End Sub
        Private Sub IXafCallbackHandler_ProcessAction(parameter As String) Implements IXafCallbackHandler.ProcessAction
            Dim indices As String() = parameter.Split(";"c)
            Dim columnIndex As Integer = Integer.Parse(indices(0))
            Dim rowIndex As Integer = Integer.Parse(indices(1))
            Dim drillDown As PivotDrillDownDataSource = DirectCast(View.Editor, ASPxPivotGridListEditor).PivotGridControl.CreateDrillDownDataSource(columnIndex, rowIndex)
            Dim name = View.ObjectTypeInfo.KeyMember.Name
            Dim keysToShow As IList = drillDown.Cast(Of PivotDrillDownDataRow)().Where(Function(row) row(name) IsNot Nothing).[Select](Function(row) row(name)).ToList()
            If keysToShow.Count > 0 Then
				Dim targetType As Type = View.ObjectTypeInfo.Type
                Dim viewId As String = Application.GetListViewId(targetType)
                Dim collectionSource As CollectionSourceBase = Application.CreateCollectionSource(Application.CreateObjectSpace(targetType), targetType, viewId)
                collectionSource.Criteria("SelectedObjects") = New InOperator(ObjectSpace.GetKeyPropertyName(targetType), keysToShow)
                Dim listView As ListView = Application.CreateListView(viewId, collectionSource, False)
                Application.ShowViewStrategy.ShowViewInPopupWindow(listView)
            End If
        End Sub
    End Class
End Namespace

File: CS/PivotGridDrillDown.Module.Web/Controllers/PivotDrillDownController.cs
using System;
using System.Collections;
using System.Linq;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.PivotGrid.Web;
using DevExpress.ExpressApp.Web;
using DevExpress.ExpressApp.Web.Templates;
using DevExpress.ExpressApp.Web.Utils;
using DevExpress.Web.ASPxPivotGrid;
using DevExpress.XtraPivotGrid;

namespace PivotGridDrillDown.Module.Web.Controllers {
    public class PivotDrillDownController : ViewController<ListView>, IXafCallbackHandler {
        private readonly string handlerId;

        protected XafCallbackManager CallbackManager {
            get { return ((ICallbackManagerHolder)WebWindow.CurrentRequestPage).CallbackManager; }
        }
        protected override void OnViewControlsCreated() {
            base.OnViewControlsCreated();

            CallbackManager.RegisterHandler(handlerId, this);

            ASPxPivotGridListEditor pivotGridListEditor = View.Editor as ASPxPivotGridListEditor;
            if(pivotGridListEditor != null) {
                ASPxPivotGrid pivotGrid = pivotGridListEditor.PivotGridControl;
                string script = CallbackManager.GetScript(handlerId, "e.ColumnIndex + ';' + e.RowIndex");
                ClientSideEventsHelper.AssignClientHandlerSafe(pivotGrid, "CellClick", "function(s, e) {" + script + "}", "pivotGrid.CellClick");
            }
        }
        public PivotDrillDownController() {
            handlerId = "PivotDrillDownHandler" + GetHashCode();
        }
        void IXafCallbackHandler.ProcessAction(string parameter) {
            string[] indices = parameter.Split(';');
            int columnIndex = Int32.Parse(indices[0]);
            int rowIndex = Int32.Parse(indices[1]);
            PivotDrillDownDataSource drillDown = ((ASPxPivotGridListEditor)View.Editor).PivotGridControl.CreateDrillDownDataSource(columnIndex, rowIndex);
            string name = View.ObjectTypeInfo.KeyMember.Name;
            IList keysToShow = drillDown.Cast<PivotDrillDownDataRow>().Where(row => row[name] != null).Select(row => row[name]).ToList();
            if(keysToShow.Count > 0) {
                Type targetType = View.ObjectTypeInfo.Type;
                string viewId = Application.GetListViewId(targetType);
                CollectionSourceBase collectionSource = Application.CreateCollectionSource(Application.CreateObjectSpace(targetType), targetType, viewId);
                collectionSource.Criteria["SelectedObjects"] = new InOperator(ObjectSpace.GetKeyPropertyName(targetType), keysToShow);
                ListView listView = Application.CreateListView(viewId, collectionSource, false);
                Application.ShowViewStrategy.ShowViewInPopupWindow(listView);
            }
        }
    }
}